### DISCUSSION TITLE
在unreal中，TypeScript编译能增加选项去压缩js代码，或者pack代码的流程么？

### DISCUSSION THREAD

**Discussion by watsonsong** [Tencent/puerts#1806] [Q&A]
目前的构建似乎是用typescript在进行编译。是否有参考流程可以构建的过程中对代码删除注释，压缩以及Pack文件。
以及编译器从tsc切换到类似swc来提高编译速度的方法？

**Comment by chexiongsheng** [comment]
没什么所谓的构建，如果你打开继承功能后，会新增一个编辑器下watch。这个watch用的也是标准的ts工程配置文件。
tsconfig.json就是标准的ts工程配置，任何ts编译器和工作量都能用。

**Comment by zhaojunmeng** [comment]
编译JS删除注释，tsc本身的配置就可以了

压缩代码是一个大话题，tsc本身都没有提供这个能力（https://github.com/microsoft/TypeScript/issues/8 这是TypeScript当时讨论的帖子，很长。。结论就是每个项目的需求可能都不同，TypeScript重复已有的轮子比如uglify-js等工具，工作量大，带来的意义却不大）

Pack文件的话，需要留意一个问题，就是modular.js里面，对于let org_require = global.require; 这部分调用
第三方bundle之后的代码，可能会引入一个global.require的函数，要看一下这个第三方生成的global.require里面，有没有生成Error对象（会在V8内部去拿Stack，造成每一次几十微秒的时间消耗）

**Reply by watsonsong** [reply]
在tsconfig.json中加了 "removeComments": true的话，不会影响CodeAnalyze本身分析注释吧？
另外压缩和Pack文件，在使用了引擎类继承的情况下，有没有一些示例的做法。因为对于TS来说支持这样的操作还挺常见的。

**Reply by zhaojunmeng** [reply]
我个人理解，CodeAnalyze本身应该是分析的ts代码而不是js代码，所以removeComments: true应该不影响CodeAnalyze

我没有实践过Pack继承引擎类的JS，第二个问题就没法给建议了

**Reply by watsonsong** [reply]
通常需要的操作是Bundle和Minification。不过tsc本身并不支持Minification，反而swc是支持Minification的。
虽然可以在最后整体打包的时候切换swc处理一遍，但日常保存后自动编译ts也是很常用的功能。在考虑CodeAnalyze.ts里面的tsc替换成swc效率能高一些，同时增加对文件Minification的功能。不知道改动起来是否困难，感觉在语法分析那边变化比较大。

**Reply by chexiongsheng** [reply]
啊。。
CodeAnalyze是编辑器用的，只在编辑器用就好了，你管它用啥，管它压不压缩？你打包时CodeAnalyze的输出不是必须用的。
CodeAnalyze的输入就是一个标准的ts工程配置（tsconfig.json），你任意的ts工具链都可以用。你爱用swc就用swc，或者任意其它ts/js工具链。

**Reply by watsonsong** [reply]
因为日常开发中，修改了TS之后输出运行都是靠CodeAnalyze产生的（它watch了TS代码并且编译输出了JS）。是希望保证日常开发运行的JS，严格和打包和整体编译构建一次的代码是完全一致的，以免产生一些不必要的BUG。
所以如果可能的话，也想让CodeAnalyze输出的js文件格式一样。当然每次整体npm run build来做一次构建也是可以的。
